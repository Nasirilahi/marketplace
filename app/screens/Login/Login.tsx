import React, {useState} from 'react';
import {
  TextInput,
  View,
  Text,
  ScrollView,
  Image,
  Keyboard,
  KeyboardAvoidingView,
  TouchableOpacity,
} from 'react-native';
import Toast from 'react-native-toast-message';
import styles from './LoginStyles';
import {validate, ValidateType} from '../../common/helpers/Validator';
import {API_URL} from '../../constants';

interface formData {
  username: string;
  password: string;
}

const LoginScreen = ({navigation}) => {
  const [userEmail, setUserEmail] = useState('');
  const [userPassword, setUserPassword] = useState('');
  const [loading, setLoading] = useState(false);
  const [errortext, setErrortext] = useState('');

  const handleSubmitPress = async () => {
    if (loading) {
      return;
    }
    if (isValid()) {
      return;
    }

    setLoading(true);
    const dataToSend = {username: userEmail, password: userPassword};
    let formBody = [];
    for (let key in dataToSend) {
      let encodedKey = encodeURIComponent(key);
      let encodedValue = encodeURIComponent(dataToSend[key]);
      formBody.push(encodedKey + '=' + encodedValue);
    }
    //
    formBody = formBody.join('&');
    // console.log('formBody', formBody)
    fetch(`${API_URL}/auth/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
      },
      body: formBody,
    })
      .then(res => res.json())
      .then(token => {
        if (token) {
          navigation.replace('AppRoutes');
        } else {
          setErrortext('Error');
          console.log('Please check your email id or password');
        }
      })
      .catch(error => {
        //Hide Loader
        setLoading(false);
        console.error(error);
      });
  };

  const isValid = (): boolean => {
    if (!validate(ValidateType.TEXT)(userEmail)) {
      Toast.show({
        type: 'error',
        text1: 'User',
        text2: 'UserName is required',
        visibilityTime: 3000,
      });
      return true;
    }
    if (!validate(ValidateType.TEXT)(userPassword)) {
      Toast.show({
        type: 'error',
        text1: 'Password',
        text2: 'Password is required',
        visibilityTime: 3000,
      });
      return true;
    }
    return false;
  };
  return (
    <View style={styles.mainBody}>
      <ScrollView
        keyboardShouldPersistTaps="handled"
        contentContainerStyle={styles.scrollContainer}>
        <KeyboardAvoidingView enabled>
          <Image
            source={require('../../common/assets/logo.png')}
            style={styles.logoImage}
          />
          <View style={styles.SectionStyle}>
            <TextInput
              style={styles.inputStyle}
              onChangeText={UserEmail => setUserEmail(UserEmail)}
              placeholder="Enter Email"
              placeholderTextColor="#8b9cb5"
              autoCapitalize="none"
              keyboardType="email-address"
              returnKeyType="next"
              underlineColorAndroid="#f000"
              blurOnSubmit={false}
            />
          </View>
          <View style={styles.SectionStyle}>
            <TextInput
              style={styles.inputStyle}
              onChangeText={UserPassword => setUserPassword(UserPassword)}
              placeholder="Enter Password"
              placeholderTextColor="#8b9cb5"
              keyboardType="default"
              onSubmitEditing={Keyboard.dismiss}
              blurOnSubmit={false}
              secureTextEntry={true}
              underlineColorAndroid="#f000"
              returnKeyType="next"
            />
          </View>
          {errortext !== '' ? (
            <Text style={styles.errorTextStyle}>{errortext}</Text>
          ) : null}
          <TouchableOpacity
            style={styles.buttonStyle}
            activeOpacity={0.5}
            onPress={handleSubmitPress}>
            <Text style={styles.buttonTextStyle}>LOGIN</Text>
          </TouchableOpacity>
        </KeyboardAvoidingView>
      </ScrollView>
    </View>
  );
};
export default LoginScreen;
