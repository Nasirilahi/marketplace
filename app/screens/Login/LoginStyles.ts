import {StyleSheet} from 'react-native';
import {Colors} from '../../constants';

const {GRO3, GRC9, GRC5, GRC1} = Colors;

const styles = StyleSheet.create({
  mainBody: {
    flex: 1,
    justifyContent: 'center',
    alignContent: 'center',
    backgroundColor: 'white',
  },
  scrollContainer: {
    flex: 1,
    justifyContent: 'center',
    alignContent: 'center',
  },
  logoImage: {
    width: '50%',
    height: 100,
    resizeMode: 'contain',
    margin: 30,
  },
  SectionStyle: {
    flexDirection: 'row',
    height: 40,
    marginTop: 10,
    marginLeft: 35,
    marginRight: 35,
    margin: 10,
  },
  buttonStyle: {
    backgroundColor: GRO3,
    borderWidth: 0,
    color: GRC9,
    borderColor: GRC5,
    height: 40,
    alignItems: 'center',
    justifyContent: 'center',
    display: 'flex',
    borderRadius: 30,
    marginLeft: 35,
    marginRight: 35,
    marginTop: 20,
    marginBottom: 25,
  },
  buttonTextStyle: {
    color: GRC9,
    paddingVertical: 10,
    fontSize: 16,
  },
  inputStyle: {
    flex: 1,
    color: GRC1,
    paddingLeft: 15,
    paddingRight: 15,
    borderWidth: 1,
    borderRadius: 30,
    borderColor: '#dadae8',
  },
  errorTextStyle: {
    color: GRO3,
    textAlign: 'center',
    fontSize: 14,
  },
});

export default styles;
