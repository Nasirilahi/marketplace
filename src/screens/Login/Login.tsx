import React, {useState, useRef, useEffect, useCallback} from 'react';
import {
  TextInput,
  View,
  Text,
  ScrollView,
  Image,
  Keyboard,
  KeyboardAvoidingView,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import Toast from 'react-native-toast-message';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Realm from 'realm';
import Token from '../../models/Token';
import styles from './LoginStyles';
import {validate, ValidateType} from '../../common/helpers/Validator';
import {API_URL} from '../../constants';

interface formData {
  username: string;
  password: string;
}

interface LoginProps {
  navigation: any;
}

const LoginScreen = ({navigation}: LoginProps) => {
  const [userName, setUserName] = useState('');
  const [userPassword, setUserPassword] = useState('');
  const [loading, setLoading] = useState(false);
  const [errortext, setErrortext] = useState('');
  const [token, setToken] = useState<Realm.Results<Token> | string>('');
  const realmRef = useRef<Realm | null>(null);

  const openRealm = useCallback(async (): Promise<void> => {
    try {
      // Open a local realm file with the schema(s) that are a part of this realm.
      const config = {
        schema: [Token],
      };
      const realm = await Realm.open(config);
      realmRef.current = realm;
      const tokenResukt: Realm.Results<Token> = realm.objects('Token');
      if (tokenResukt?.length) {
        setToken(token);
      }
    } catch (err) {
      console.error('Error opening realm: ', err.message);
    }
  }, [realmRef, setToken, token]);

  const closeRealm = useCallback((): void => {
    const realm = realmRef.current;
    realm?.close();
    realmRef.current = null;
    setToken('');
  }, [realmRef]);

  useEffect(() => {
    openRealm();

    // Return a cleanup callback to close the realm to prevent memory leaks
    return closeRealm;
  }, [openRealm, closeRealm]);

  const handleSubmitPress = async () => {
    if (loading) {
      return;
    }
    if (isValid()) {
      return;
    }

    setLoading(true);
    const dataToSend = {username: userName, password: userPassword};
    let formBody = [];
    for (let key in dataToSend) {
      let encodedKey = encodeURIComponent(key);
      let encodedValue = encodeURIComponent(dataToSend[key]);
      formBody.push(encodedKey + '=' + encodedValue);
    }
    formBody = formBody.join('&');
    fetch(`${API_URL}/auth/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
      },
      body: formBody,
    })
      .then(res => res.json())
      .then(reponse => {
        if (reponse?.token) {
          const realm = realmRef.current;
          realm?.write(() => {
            realm?.create('Token', Token.generate(reponse.token));
          });
          AsyncStorage.setItem('token', reponse.token);
          navigation.replace('AppRoutes');
        } else {
          setErrortext('Error');
          console.log('Please check your email id or password');
        }
      })
      .catch(error => {
        //Hide Loader
        setLoading(false);
      });
  };

  const isValid = (): boolean => {
    if (!validate(ValidateType.TEXT)(userName)) {
      Toast.show({
        type: 'error',
        text1: 'User',
        text2: 'UserName is required',
        visibilityTime: 3000,
      });
      return true;
    }
    if (!validate(ValidateType.TEXT)(userPassword)) {
      Toast.show({
        type: 'error',
        text1: 'Password',
        text2: 'Password is required',
        visibilityTime: 3000,
      });
      return true;
    }
    return false;
  };
  return (
    <View style={styles.mainBody}>
      <ScrollView
        keyboardShouldPersistTaps="handled"
        contentContainerStyle={styles.scrollContainer}>
        <KeyboardAvoidingView enabled>
          <Image
            source={require('../../common/assets/logo.png')}
            style={styles.logoImage}
          />
          <View style={styles.SectionStyle}>
            <TextInput
              style={styles.inputStyle}
              onChangeText={useName => setUserName(useName)}
              placeholder="Enter User Name"
              placeholderTextColor="#8b9cb5"
              autoCapitalize="none"
              keyboardType="email-address"
              returnKeyType="next"
              underlineColorAndroid="#f000"
              blurOnSubmit={false}
              testID="usernameInput"
            />
          </View>
          <View style={styles.SectionStyle}>
            <TextInput
              style={styles.inputStyle}
              onChangeText={UserPassword => setUserPassword(UserPassword)}
              placeholder="Enter Password"
              placeholderTextColor="#8b9cb5"
              keyboardType="default"
              onSubmitEditing={Keyboard.dismiss}
              blurOnSubmit={false}
              secureTextEntry={true}
              underlineColorAndroid="#f000"
              returnKeyType="next"
              testID="passwordInput"
            />
          </View>
          {errortext !== '' ? (
            <Text style={styles.errorTextStyle}>{errortext}</Text>
          ) : null}
          <TouchableOpacity
            style={styles.buttonStyle}
            activeOpacity={0.5}
            onPress={handleSubmitPress}>
            {loading ? (
              <ActivityIndicator size="small" color={'white'} />
            ) : (
              <Text style={styles.buttonTextStyle}>LOGIN</Text>
            )}
          </TouchableOpacity>
        </KeyboardAvoidingView>
      </ScrollView>
    </View>
  );
};
export default LoginScreen;
