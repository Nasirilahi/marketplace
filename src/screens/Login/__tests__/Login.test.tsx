import React from 'react';
import {render, fireEvent, act} from '@testing-library/react-native';
import Login from '..';

const flushMicrotasksQueue = () =>
  new Promise(resolve => setImmediate(resolve));

describe('Login', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  it('renders default elements', () => {
    const navigate = jest.fn();
    const {getAllByText, getByPlaceholderText} = render(
      <Login navigation={{navigate: navigate}} />,
    );
    expect(getAllByText('LOGIN').length).toBe(1);
    getByPlaceholderText('Enter User Name');
    getByPlaceholderText('Enter Password');
  });
  // it('shows invalid input messages', async() => {
  //   const navigate = jest.fn();
  //   const {getByTestId, getByText} = render(<Login navigation={{navigate: navigate}}  />);

  //   fireEvent.press(getByText('LOGIN'));
  //   expect(await getByText('User').length).toBe(1);
  //   getByText('Password is required');
  // });
  // it('handles valid input submission', async () => {
  //   fetch.mockResponseOnce(JSON.stringify({token: 'sdfxsdfsd'}));
  //   const pushMock = jest.fn();

  //   const navigate = jest.fn();
  //   const {getByTestId, getByText} = render(
  //     <Login navigation={{navigate: navigate}} />,
  //   );
  //   fireEvent.changeText(getByTestId('usernameInput'), 'example');
  //   fireEvent.changeText(getByTestId('passwordInput'), 'asdf');
  //   fireEvent.press(getByText('LOGIN'));

  //   expect(fetch.mock.calls).toMatchSnapshot();
  //   await act(flushMicrotasksQueue);
  //   expect(pushMock).toBeCalled();
  // });
});
