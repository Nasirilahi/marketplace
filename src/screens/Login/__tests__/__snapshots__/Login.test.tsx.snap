// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login handles valid input submission 1`] = `
Array [
  Array [
    "https://fakestoreapi.com/auth/login",
    Object {
      "body": "username=example&password=asdf",
      "headers": Object {
        "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
      },
      "method": "POST",
    },
  ],
]
`;
